File: PyAlDftr/AlDftr.py
"""
ALDftr: Wiki-Based Knowledge Organizer
"""
__author__ = "Mazen A. Melibari"
__email__ = "mazen@mazen.ws"
__license__ = "MPL"
__version__ = "0.1"

from flask import Flask, request, render_template, jsonify, redirect, url_for
import os
import sys
import re
import json
import codecs
import webbrowser

LOCAL_SERVER_PORT = 8001
ROOT_FOLDER = os.path.dirname(sys.argv[0])
DATA_FOLDER = unicode(os.path.join(ROOT_FOLDER, '../data')) # use unicode so that os.listdir() returns filenames in unicode!
METADATA_SEPARATOR = '#####-----|+|-|-|+|-----#####' # new lines added before and after the separator


#------------------------------------------------------------
# AlDftr's Main Class
#------------------------------------------------------------
class AlDftr:
    def __init__(self, data_folder, metadata_separator):
        self.data_folder = data_folder
        self.metadata_separator = metadata_separator

    def sanitize_page_name(self, page_name):
        """
        Returns a sanitized page name
        """
        # replace any nonalphanumeric(english and arabic) char or '_' by a dash
        #r'[^0-9a-zA-Z\u0600-\u06FF\_]+',
        regexp = re.compile(r'[^\w\d\:]+', re.UNICODE)
        return regexp.sub('-', page_name)

    def get_file_path(self, page_name):
        """
        Returns the full path of the page file
        """
        page_path_parts = page_name.split(':')
        page_path_parts[-1] = page_path_parts[-1] + '.txt'
        file_path = os.path.join(self.data_folder, *page_path_parts)
        return file_path

    def get_page(self, page_name):
        """
        Returns a tuple with  the content and metadata of the page
        """
        file_path = self.get_file_path(page_name)
        with codecs.open(file_path, 'r', 'utf-8') as file:
            file_content = file.read()

        (page_content, page_metadata) = self.parse_file_content(file_content)

        return (page_content, page_metadata)

    def save_page(self, page_name, page_content, page_metadata):
        file_path = self.get_file_path(page_name)

        #create namespace if not existed
        if not self.is_namespace_exists(file_path):
            os.makedirs(os.path.dirname(file_path))

        # append metadata to the end of the page
        file_content = page_content + '\n' + self.metadata_separator + '\n' + page_metadata
        with codecs.open(file_path, 'w', 'utf-8-sig') as file:
            file.write(file_content)

    def delete_page(self, page_name):
        file_path = self.get_file_path(page_name)
        os.remove(file_path)

    def is_page_exists(self, page_name):
        file_path = self.get_file_path(page_name)
        return os.path.exists(file_path)

    def is_namespace_exists(self, page_full_path):
        namespace_path = os.path.dirname(page_full_path)
        return os.path.exists(namespace_path)

    def get_all_pages(self):
        return os.listdir(self.data_folder)

    def parse_file_content(self, file_content):
        """
        Given a full page contents, this function separates the content and metadata and returns them
        """
        # prepare metadata sepration regular expression
        regexp = re.compile(r'\n' + re.escape(self.metadata_separator) + '.*', re.S)
        find_metadata = regexp.findall(file_content)
        if len(find_metadata):
            page_metadata = find_metadata[0]
            page_metadata = page_metadata.replace(self.metadata_separator, '')
            page_metadata = json.loads(page_metadata)
        else:
            page_metadata = {}

        # remove metadata from the content
        page_content = regexp.sub('', file_content)

        return (page_content, page_metadata)

    def dftr_format_to_html(self, text):
        """
        dftrFormat is a lightweight markup language that's inspired by Markdown, Wikiformat, and reStructuredText.
        It's designed specifically to be easily written in LTR and RTL languages.

        Current Specification (v0.001):
        \n                  : new line
        [[page-name]]       : local link to page-name
        **string**          : bold
        -----[-*]           : hr line
        #string             : h1
        ##string            : h2
        ###string           : h3
        ####string          : h4
        #####string         : h5
        ######string         : h6
        """

        text = text.replace('\r', '')

        local_link_regexp = re.compile('\[\[([^\]]*)\]\]')
        for page_name in local_link_regexp.findall(text):
            page_name = page_name.strip()
            page_link = '<a href="%s">%s</a>' % (url_for('view', page_name = page_name), page_name)
            text = text.replace('[[%s]]' % page_name, page_link)

        bold_regexp = re.compile(r'\*\*(.*?)\*\*')
        text = bold_regexp.sub(r'<b>\1</b>', text)

        hr_regexp = re.compile(r'^\-\-\-\-\-\-*$', re.M)
        text = hr_regexp.sub(r'<hr>', text)

        h6_regexp = re.compile(r'^\#\#\#\#\#\#(.*)$', re.M)
        text = h6_regexp.sub(r'<h6>\1</h6>', text)

        h5_regexp = re.compile(r'^\#\#\#\#\#(.*)$', re.M)
        text = h5_regexp.sub(r'<h5>\1</h5>', text)

        h4_regexp = re.compile(r'^\#\#\#\#(.*)$', re.M)
        text = h4_regexp.sub(r'<h4>\1</h4>', text)

        h3_regexp = re.compile(r'^\#\#\#(.*)$', re.M)
        text = h3_regexp.sub(r'<h3>\1</h3>', text)

        h2_regexp = re.compile(r'^\#\#(.*)$', re.M)
        text = h2_regexp.sub(r'<h2>\1</h2>', text)

        h1_regexp = re.compile(r'^\#(.*)$', re.M)
        text = h1_regexp.sub(r'<h1>\1</h1>', text)

        text = text.replace('\n', '<br>')

        return text

#------------------------------------------------------------
# Frontend
#------------------------------------------------------------

# Prepare global objects
aldftr = AlDftr(DATA_FOLDER, METADATA_SEPARATOR)
app = Flask(__name__)

# this function injects variables into all the templates.
# The 'endpoint' contains the name of the current action [index, view, or edit]
# we need it to change the look of the layout according to the current action
@app.context_processor
def inject():
    return dict(endpoint=request.endpoint, __version__=__version__)

# a wrapper around 'dftr_format_to_html' to convert from dftrFormat to html
@app.template_filter('dftr')
def dftr(s):
    return aldftr.dftr_format_to_html(s)


###----------------------------------------------------------
### Routes
###----------------------------------------------------------

@app.route('/')
def index():
    return redirect(url_for('view', page_name = 'main'))

@app.route('/all_pages')
def all_pages():
    pages = aldftr.get_all_pages()
    print(pages)
    return render_template('all_pages.html', pages = pages)

@app.route('/view/<page_name>')
def view(page_name):
    page_name = aldftr.sanitize_page_name(page_name)
    if aldftr.is_page_exists(page_name):
        (page_content, page_metadata) = aldftr.get_page(page_name)
        return render_template('view.html', page_name = page_name, page_content = page_content, page_metadata = page_metadata)
    else:
        return redirect(url_for('edit', page_name = page_name))

@app.route('/edit/<page_name>', methods=['POST', 'GET'])
def edit(page_name):
    if request.method == 'POST':
        page_name = aldftr.sanitize_page_name(request.form['page_name'])
        page_content = request.form['page_content']
        page_metadata = request.form['page_metadata']
        aldftr.save_page(page_name, page_content, page_metadata)

        return redirect(url_for('view', page_name = page_name))

    else:
        page_name = aldftr.sanitize_page_name(page_name)
        page_content = ''
        page_metadata = {}

        if aldftr.is_page_exists(page_name):
            (page_content, page_metadata) = aldftr.get_page(page_name)

        return render_template('edit.html', page_name = page_name, page_content = page_content, page_metadata = page_metadata)

@app.route('/delete/<page_name>')
def delete(page_name):
    page_name = aldftr.sanitize_page_name(page_name)
    aldftr.delete_page(page_name)
    return redirect(url_for('index'))


if __name__ == '__main__':
    print("#"*50)
    print("Aldftr v%s" % __version__)
    print("Please keep this window open as long as you want to access AlDftr")
    print("#"*50)

    app.debug = True
    webbrowser.open("http://localhost:8001")
    app.run(port=LOCAL_SERVER_PORT)

File: GoAlDftr/AlDftr.go
/*
 * ALDftr
 * Wiki-Based Knowledge Organizer
 *
 * author: "Mazen A. Melibari"
 * email: "mazen@mazen.ws"
 * license: "MPL"
 * version: "0.1"
 */
package main

import (
	"strings"
	"bytes"
	"regexp"
	"path/filepath"
	"io"
	"io/ioutil"
	"net/http"
	"html/template"
	"encoding/json"
	"os"
	"os/exec"
	"runtime"
	"fmt"
)

var ROOT_PATH = filepath.Dir(os.Args[0])
var __VERSION__ = "0.2"
var STATIC_PATH = filepath.Join(ROOT_PATH, "etc", "static")
var TEMPLATE_PATH = filepath.Join(ROOT_PATH, "etc", "templates")
var METADATA_SEPARATOR = "#####-----|+|-|-|+|-----#####"
var DATA_FOLDER =  filepath.Join(ROOT_PATH, "data") + string(filepath.Separator) // trailing slash is mandatory
var LOCAL_SERVER_PORT = "5000"

type Page struct {
	Name []byte
	Content []byte
	Metadata map[string]string
}

func (p *Page) sanitize_page_name() {
	re := regexp.MustCompile("[^\\p{Arabic}\\w\\d\\:]+")
	p.Name = re.ReplaceAll(p.Name, []byte("-"))
}

func (p *Page) get_file_path() string {
	file_path_parts := strings.Split(string(p.Name), ":")
	file_path_parts[len(file_path_parts)-1] = file_path_parts[len(file_path_parts)-1] + ".txt"

	file_internal_path := filepath.Join(file_path_parts...)
	file_path := filepath.Join(DATA_FOLDER, file_internal_path)
	return file_path
}

func (p *Page) load() error {
	file_path := p.get_file_path()
	content, err := ioutil.ReadFile(file_path)
	if err == nil {
		p.Content = content
		p.parse_file_content()
	}
	return err
}

func (p *Page) save() error {
	file_full_path := p.get_file_path()
	file_dir_path := filepath.Dir(file_full_path)
	os.MkdirAll(file_dir_path, 0600)

	metadata_json, _ := json.Marshal(p.Metadata)
	content := string(p.Content) + "\n" + METADATA_SEPARATOR + "\n" + string(metadata_json)
	return ioutil.WriteFile(file_full_path, []byte(content), 0600)
}

func (p *Page) delete() error {
	file_full_path := p.get_file_path()
	return os.Remove(file_full_path)
}

func (p *Page) parse_file_content() {
	separator := "\n" + METADATA_SEPARATOR + "\n"
	re := regexp.MustCompile(regexp.QuoteMeta(separator) + ".*")

	found_metadata := re.FindAll(p.Content, -1)
	if found_metadata != nil {
		metadata_text := strings.Replace(string(found_metadata[0]), separator, "", 1)
		p.Metadata = make(map[string]string)
		json.Unmarshal([]byte(metadata_text), &p.Metadata)
	}

	p.Content = re.ReplaceAll(p.Content, []byte(""))
}

func get_all_pages(data_folder_path string) []string {
	pages := make([]string, 0, 0)
	folder_separator := string(filepath.Separator)

	visit := func(path string, info os.FileInfo, err error) error {
		if info.IsDir() {
			return nil
		}

		page_path := strings.Replace(path, DATA_FOLDER, "", 1)
		page_name := strings.Replace(page_path, folder_separator, ":", -1)

		if page_name != "" {
			pages = append(pages, page_name[:len(page_name)-4])
		}

		return nil

	}

	filepath.Walk(data_folder_path, visit)
	return pages
}

func dftr_format_to_html(txt []byte) []byte {
	// remove \r
	re := regexp.MustCompile(`\r`)
	txt = re.ReplaceAll(txt, []byte(""))

	// local links
	re = regexp.MustCompile(`\[\[([^\]]*)\]\]`)
	txt = re.ReplaceAll(txt, []byte("<a href=\"../view/$1\">$1</a>"))

	// bold
	re = regexp.MustCompile(`\*\*(.*?)\*\*`)
	txt = re.ReplaceAll(txt, []byte("<b>$1</b>"))

	// hr
	re = regexp.MustCompile(`(?m)^\-\-\-\-\-\-*$`)
	txt = re.ReplaceAll(txt, []byte("<hr>"))

	// h6
	re = regexp.MustCompile(`(?m)^\#\#\#\#\#\#(.*)$`)
	txt = re.ReplaceAll(txt, []byte("<h6>$1</h6>"))

	// h5
	re = regexp.MustCompile(`(?m)^\#\#\#\#\#(.*)$`)
	txt = re.ReplaceAll(txt, []byte("<h5>$1</h5>"))

	// h4
	re = regexp.MustCompile(`(?m)^\#\#\#\#(.*)$`)
	txt = re.ReplaceAll(txt, []byte("<h4>$1</h4>"))

	// h3
	re = regexp.MustCompile(`(?m)^\#\#\#(.*)$`)
	txt = re.ReplaceAll(txt, []byte("<h3>$1</h3>"))

	// h2
	re = regexp.MustCompile(`(?m)^\#\#(.*)$`)
	txt = re.ReplaceAll(txt, []byte("<h2>$1</h2>"))

	// h1
	re = regexp.MustCompile(`(?m)^\#(.*)$`)
	txt = re.ReplaceAll(txt, []byte("<h1>$1</h1>"))

	// br
	re = regexp.MustCompile(`\n`)
	txt = re.ReplaceAll(txt, []byte("<br>"))

	return txt
}

func render_template(wr io.Writer, template_name string, data interface{}) {
	var template_results bytes.Buffer
	full_template_path := filepath.Join(TEMPLATE_PATH, template_name + ".html")
	tmpl, _ := template.ParseFiles(full_template_path)
	tmpl.Execute(&template_results, data)

	var page_name string
	value_of_data := data.(map[string]interface{})
	if val, exists := value_of_data["page_name"]; exists {
		page_name = string(val.([]byte))
	} else {
		page_name = ""
	}

	layout_data := map[string]interface {} {
		"Body": template.HTML(template_results.Bytes()),
		"page_name": page_name,
		"__VERSION__": __VERSION__,
	}
	layout_tmpl, _ := template.ParseFiles(filepath.Join(TEMPLATE_PATH, "layout.html"))
	layout_tmpl.Execute(wr, layout_data)
}


func get_page_name_from_url(r *http.Request, action_name string) string {
	page_name := r.URL.Path[len("/" + action_name + "/"):]
	return page_name
}


func index(w http.ResponseWriter, r *http.Request) {
	http.Redirect(w, r, "../view/Main", http.StatusFound)
}

func view(w http.ResponseWriter, r *http.Request) {
	page_name := get_page_name_from_url(r, "view")

	page := Page{Name: []byte(page_name)}
	page.sanitize_page_name()

	if page.Name == nil {
		http.Redirect(w, r, "../view/Main", http.StatusNotFound)
		return
	}

	err := page.load()

	if err == nil {
		page.Content = dftr_format_to_html(page.Content)

		// Capitalize first char
		data := map[string]interface {} {
			"page_name": page.Name,
			"page_content": template.HTML(page.Content),
			"page_metadata": page.Metadata,
		}
		render_template(w, "view", data)
	} else {
		http.Redirect(w, r, "../edit/" + string(page.Name), http.StatusFound)
		return
	}
}

func edit(w http.ResponseWriter, r *http.Request) {
	page_name := get_page_name_from_url(r, "edit")

	page := Page{Name: []byte(page_name)}
	page.sanitize_page_name()

	if page.Name == nil {
		http.Redirect(w, r, "../view/Main", http.StatusNotFound)
		return
	}

	page.load()

	metadata, _ := json.Marshal(page.Metadata)

	data := map[string]interface {} {
		"page_name": page.Name,
		"page_content": page.Content,
		"page_metadata": metadata,
	}

	render_template(w, "edit", data)
}

func save(w http.ResponseWriter, r *http.Request) {
	page_name := []byte(r.FormValue("page_name"))
	page_content := []byte(r.FormValue("page_content"))
	page_metadata := []byte(r.FormValue("page_metadata"))

	page := Page{Name: page_name, Content: page_content}
	json.Unmarshal(page_metadata, &page.Metadata)
	page.sanitize_page_name()

	if page.Name == nil {
		http.Redirect(w, r, "../view/Main", http.StatusNotFound)
		return
	}

	page.save()

	http.Redirect(w, r, "../view/" + string(page.Name), http.StatusFound)
}

func delete(w http.ResponseWriter, r *http.Request) {
	page_name := get_page_name_from_url(r, "delete")
	page := Page{Name: []byte(page_name)}
	page.delete()
	http.Redirect(w, r, "../view/Main", http.StatusFound)
}

func all_pages(w http.ResponseWriter, r *http.Request) {
	pages := get_all_pages(DATA_FOLDER)
	data := map[string]interface {} {
		"pages": pages,
	}
	render_template(w, "all_pages", data)
}

func main() {
	server_addr := "localhost:" + LOCAL_SERVER_PORT
	full_server_addr := "http://" + server_addr

	fmt.Println(DATA_FOLDER)

	fmt.Println("AlDftr v" + __VERSION__)
	fmt.Println("Please keep this window open")
	fmt.Println("-----------------------------")
	fmt.Println("AlDftr running on:" + full_server_addr)
	get_all_pages(DATA_FOLDER)

	switch runtime.GOOS {
	case "windows":
		exec.Command("cmd", "/c", "start", full_server_addr).Start()
	case "darwin":
		exec.Command("open", full_server_addr).Start()
	case "linux":
		exec.Command("xdg-open", full_server_addr).Start()
	}

	http.HandleFunc("/", index)
	http.Handle("/static/", http.StripPrefix("/static", http.FileServer(http.Dir(STATIC_PATH))))
	http.HandleFunc("/view/", view)
	http.HandleFunc("/edit/", edit)
	http.HandleFunc("/save/", save)
	http.HandleFunc("/delete/", delete)
	http.HandleFunc("/all_pages/", all_pages)

	http.ListenAndServe(server_addr, nil)
}
